### ISSUE TITLE
clang16 false positive error "binds to a temporary"

### ISSUE THREAD

**Issue by kelbon** [llvm/llvm-project#64934]
https://godbolt.org/z/88a9W7deY

```C++
dd::generator<int> g2() {
// well formed in clang15, ill formed in clang 16
// error: reference member 'rng' binds to a temporary object
// whose lifetime would be shorter
// than the lifetime of the constructed object
    co_yield dd::elements_of(g1());
}
```

**Comment by philnik777** [comment]
Reduced test-case: https://godbolt.org/z/89eajaeG6

**Comment by kelbon** [comment]
> Reduced test-case: https://godbolt.org/z/89eajaeG6

https://godbolt.org/z/fb1s5evsh
